File: slp/backend-nestjs/src/modules/avatar/avatar.controller.ts
import {
  Controller,
  Post,
  Delete,
  UseGuards,
  UseInterceptors,
  UploadedFile,
  HttpCode,
  HttpStatus,
  BadRequestException,
  UnsupportedMediaTypeException,
  PayloadTooLargeException,
  NotFoundException,
  Logger,
  Request,
} from "@nestjs/common";
import { FileInterceptor } from "@nestjs/platform-express";
import { SessionGuard } from "../session/session.guard";
import { UserRepository } from "../user/user.repository";
import { FileStorageService } from "../file-storage/file-storage.service";
import { ConfigService } from "@nestjs/config";
import { AvatarResponseDto } from "./dto/avatar-response.dto";

@Controller("api/avatar")
@UseGuards(SessionGuard)
export class AvatarController {
  private readonly logger = new Logger(AvatarController.name);
  private readonly allowedMime = new Set(["image/jpeg", "image/png"]);
  private readonly maxBytes = 2 * 1024 * 1024; // 2 MB
  private readonly frontendBaseUrl: string;

  constructor(
    private readonly userRepo: UserRepository,
    private readonly fileStorage: FileStorageService,
    private readonly configService: ConfigService,
  ) {
    this.frontendBaseUrl =
      this.configService.get<string>("fileStorage.baseUrlFrontend") || "";
  }

  @Post()
  @UseInterceptors(FileInterceptor("file"))
  async uploadAvatar(
    @UploadedFile() file: Express.Multer.File,
    @Request() req,
  ): Promise<AvatarResponseDto> {
    const userId = req.user.id;

    // Validation
    if (!file) {
      throw new BadRequestException("No file provided.");
    }
    if (!this.allowedMime.has(file.mimetype)) {
      throw new UnsupportedMediaTypeException(
        "Only JPEG and PNG images are allowed.",
      );
    }
    if (file.size > this.maxBytes) {
      throw new PayloadTooLargeException("File exceeds the 2 MB limit.");
    }

    // Get user
    const user = await this.userRepo.getById(userId);
    if (!user) {
      throw new NotFoundException("User not found.");
    }

    // Delete old avatar (best-effort)
    if (user.avatarFilename) {
      try {
        await this.fileStorage.deleteFileAsync(user.avatarFilename);
      } catch (err) {
        this.logger.warn(
          `Could not delete previous avatar for user ${userId}`,
          err,
        );
      }
    }

    // Upload new file
    let newFilename: string;
    try {
      newFilename = await this.fileStorage.uploadAvatarAsync(
        file.buffer,
        file.mimetype,
        file.originalname,
      );
    } catch (err) {
      this.logger.error(`Avatar upload failed for user ${userId}`, err);
      throw err; // already handled by service
    }

    // Update user
    user.avatarFilename = newFilename;
    user.updatedAt = new Date();
    await this.userRepo.update(user);

    // Build public URL
    const avatarUrl = `${this.frontendBaseUrl.endsWith("/") ? this.frontendBaseUrl.slice(0, -1) : this.frontendBaseUrl}/files/${newFilename}`;
    return { avatarUrl };
  }

  @Delete()
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteAvatar(@Request() req): Promise<void> {
    const userId = req.user.id;
    const user = await this.userRepo.getById(userId);
    if (!user) {
      throw new NotFoundException("User not found.");
    }

    if (user.avatarFilename) {
      try {
        await this.fileStorage.deleteFileAsync(user.avatarFilename);
      } catch (err) {
        this.logger.warn(`Could not delete avatar for user ${userId}`, err);
      }

      user.avatarFilename = undefined;
      user.updatedAt = new Date();
      await this.userRepo.update(user);
    }
    // NoContent
  }
}

File: slp/backend-dotnet/Features/Avatar/AvatarController.cs
using backend_dotnet.Features.FileStorage;
using backend_dotnet.Features.User;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;
using System.Security.Claims;

namespace backend_dotnet.Features.Avatar;

/// <summary>
/// Handles avatar upload and removal for the authenticated user.
/// The database stores only the bare filename; the public URL is computed here.
/// Route: /api/avatar
/// </summary>
[ApiController]
[Route("api/avatar")]
[Authorize]
public sealed class AvatarController : ControllerBase
{
    private static readonly HashSet<string> AllowedMime =
        new(StringComparer.OrdinalIgnoreCase) { "image/jpeg", "image/png" };

    private const long MaxBytes = 2 * 1024 * 1024; // 2 MB

    private readonly IFileStorageClient _storage;
    private readonly IUserRepository _users;
    private readonly FileStorageSettings _settings;
    private readonly ILogger<AvatarController> _logger;

    public AvatarController(
        IFileStorageClient storage,
        IUserRepository users,
        IOptions<FileStorageSettings> settings,
        ILogger<AvatarController> logger)
    {
        _storage  = storage;
        _users    = users;
        _settings = settings.Value;
        _logger   = logger;
    }

    // ── POST /api/avatar ──────────────────────────────────────────────────────

    /// <summary>Upload or replace the authenticated user's avatar.</summary>
    [HttpPost]
    [Consumes("multipart/form-data")]
    public async Task<IActionResult> UploadAvatar(IFormFile file)
    {
        var userId = GetCurrentUserId();
        if (userId == null) return Unauthorized();

        // --- Validate ---
        if (file == null || file.Length == 0)
            return BadRequest(new { message = "No file provided." });

        if (!AllowedMime.Contains(file.ContentType))
            return StatusCode(415, new { message = "Only JPEG and PNG images are allowed." });

        if (file.Length > MaxBytes)
            return StatusCode(413, new { message = "File exceeds the 2 MB limit." });

        // --- Load user ---
        var user = await _users.GetByIdAsync(userId.Value);
        if (user == null) return NotFound();

        // --- Delete old avatar file (best-effort) ---
        if (!string.IsNullOrWhiteSpace(user.AvatarFilename))
        {
            try { await _storage.DeleteFileAsync(user.AvatarFilename); }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Could not delete previous avatar for user {UserId}", userId);
            }
        }

        // --- Upload new file; service returns just the filename ---
        await using var ms = new MemoryStream();
        await file.CopyToAsync(ms);
        var bytes = ms.ToArray();

        string newFilename;
        try
        {
            newFilename = await _storage.UploadAvatarAsync(bytes, file.ContentType, file.FileName);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Avatar upload failed for user {UserId}", userId);
            return StatusCode(502, new { message = "File storage service error. Please try again." });
        }

        // --- Persist bare filename only ---
        user.AvatarFilename = newFilename;
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user);

        // --- Return the constructed public URL to the client ---
        return Ok(new { avatarUrl = newFilename });
    }

    // ── DELETE /api/avatar ────────────────────────────────────────────────────

    /// <summary>Remove the authenticated user's avatar.</summary>
    [HttpDelete]
    public async Task<IActionResult> DeleteAvatar()
    {
        var userId = GetCurrentUserId();
        if (userId == null) return Unauthorized();

        var user = await _users.GetByIdAsync(userId.Value);
        if (user == null) return NotFound();

        if (!string.IsNullOrWhiteSpace(user.AvatarFilename))
        {
            try { await _storage.DeleteFileAsync(user.AvatarFilename); }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Could not delete avatar for user {UserId}", userId);
            }

            user.AvatarFilename = null;
            user.UpdatedAt = DateTime.UtcNow;
            await _users.UpdateAsync(user);
        }

        return NoContent();
    }

    // ── Private helpers ───────────────────────────────────────────────────────

    private int? GetCurrentUserId()
    {
        var raw = User.FindFirstValue(ClaimTypes.NameIdentifier);
        return int.TryParse(raw, out var id) ? id : null;
    }
}